apps/cowswap-frontend/src/modules/bridge/pure/ProtocolIcons/index.tsx

```diff
@@ -10,63 +12,215 @@ export interface ProtocolIconsProps {
-          overrideHoverColor={`var(${UI.COLOR_BLUE_300_PRIMARY})`}
-        />
-      </ProtocolIcon>
-    )
+interface MaskCalculationResult {
+  firstMaskConfig: MaskConfig | undefined
+  secondMaskConfig: MaskConfig | undefined
+  isFirstOnTop: boolean
+  isSecondOnTop: boolean
+}
+
+// Custom hook for hover state management
+function useIconHoverState(): IconHoverState {
+  const [hoveredIcon, setHoveredIcon] = useState<'first' | 'second' | null>(null)
+
+  const handleFirstIconMouseEnter = (): void => setHoveredIcon('first')
+  const handleSecondIconMouseEnter = (): void => setHoveredIcon('second')
+  const handleMouseLeave = (): void => setHoveredIcon(null)
+
+  return {
+    hoveredIcon,
+    handleFirstIconMouseEnter,
+    handleSecondIconMouseEnter,
+    handleMouseLeave,
   }
+}
 
-  if (showOnlySecond) {
-    return (
-      <ProtocolIcon title={secondProtocol.name} size={iconSize}>
-        <img src={secondProtocol.logoUrl} width={logoHeight} height={logoHeight} alt={secondProtocol.name} />
-      </ProtocolIcon>
-    )
+// Utility function for mask configuration
+const createMask = (
+  baseCx: number,
+  offsetMultiplier: number,
+  distance: number,
+  centerY: number,
+  innerR: number,
+  outerR: number,
+): MaskConfig => ({
+  cx: baseCx + offsetMultiplier * distance,
+  cy: centerY,
+  innerR: innerR,
+  outerR: outerR,
+})
+
+// Custom hook for mask calculations. Memoized only for costly computation.
+function useMaskCalculations(
+  currentDisplaySize: number,
+  hoveredIcon: 'first' | 'second' | null,
+): MaskCalculationResult {
+  return useMemo(() => {
+    const cutThickness = getBorderWidth(currentDisplaySize)
+    const radiusOfTopIcon = currentDisplaySize / 2
+    const maskOuterRadius = radiusOfTopIcon + cutThickness
+    const maskCenterY = currentDisplaySize / 2
+    const iconCenterBaseX = currentDisplaySize / 2
+    const distanceBetweenCenters = currentDisplaySize * (1 - STACKED_ICON_OVERLAP_RATIO)
+
+    const isFirstHovered = hoveredIcon === 'first'
+    const isSecondHovered = hoveredIcon === 'second'
+    const isFirstOnTop = isFirstHovered
+    const isSecondOnTop = isSecondHovered || (!isFirstHovered && !isSecondHovered)
+
+    const firstMaskConfig = !isFirstOnTop
+      ? createMask(iconCenterBaseX, 1, distanceBetweenCenters, maskCenterY, radiusOfTopIcon, maskOuterRadius)
+      : undefined
+    const secondMaskConfig = !isSecondOnTop
+      ? createMask(iconCenterBaseX, -1, distanceBetweenCenters, maskCenterY, radiusOfTopIcon, maskOuterRadius)
+      : undefined
+
+    return {
+      firstMaskConfig,
+      secondMaskConfig,
+      isFirstOnTop,
+      isSecondOnTop,
+    }
+  }, [currentDisplaySize, hoveredIcon])
+}
+
+// Single Protocol Icon Component
+interface SingleProtocolIconProps {
+  showOnlyFirst?: boolean
+  secondProtocol: BridgeProviderInfo
+  currentDisplaySize: number
+  currentLogoHeight: number
+}
+
+const SingleProtocolIcon = memo(function SingleProtocolIcon({
```

Comment:
Is `memo()` needed here?

Reply:
To consider why `memo()` could be needed here:

The `SingleProtocolIcon` component receives stable calculated props (`currentDisplaySize`, `currentLogoHeight`) that often don't change between parent re-renders. Memoizing prevents unnecessary re-rendering of DOM elements and image loading when props remain the same, following the guideline to wrap pure components with `memo()` when props rarely change.